packages/server/src/loop.test.ts

```diff
@@ -0,0 +1,135 @@
+  });
+
+  beforeEach(() => {
+    // tell vitest we use mocked time
+    vi.useFakeTimers();
+  });
+
+  afterEach(() => {
+    // restoring date after each test run
+    vi.useRealTimers();
+  });
+
+  it('should loop and charge for open invoices', async () => {
+    // given
+    const testData = getFixtures();
+
+    const customer = testData.customer;
+    customer.activePaymentMethod = testData.paymentMethod;
+
+    const paymentProvider = getPaymentProvider(testData.project);
+    if (!paymentProvider) {
+      throw new Error('Payment provider not configured');
+    }
+    await paymentProvider.createCustomer(customer);
+
+    const db = {
+      invoices: new Map<string, Invoice>(),
+      payments: new Map<string, Payment>(),
+      customers: new Map<string, Customer>(),
+    };
+
+    db.invoices.set(testData.invoice._id, testData.invoice);
+
+    vi.spyOn(databaseExports, 'database', 'get').mockReturnValue({
+      em: {
+        persistAndFlush: async (_items: unknown[] | unknown) => {
+          const items = Array.isArray(_items) ? _items : [_items];
+
+          for (const item of items) {
+            // console.log('set', item._id, item.constructor.name);
+            if (item instanceof Invoice) {
+              db.invoices.set(item._id, item);
+            } else if (item instanceof Payment) {
+              db.payments.set(item._id, item);
+            } else if (item instanceof Customer) {
+              db.customers.set(item._id, item);
+            }
+          }
+
+          return Promise.resolve();
+        },
+        populate() {
+          return Promise.resolve();
+        },
+      },
+      invoices: {
+        find() {
+          return Promise.resolve(Array.from(db.invoices.values()));
+        },
+      },
+    } as unknown as databaseExports.Database);
+
+    const invoiceDate = dayjs(testData.invoice.date);
+    vi.setSystemTime(invoiceDate.add(1, 'day').toDate());
+
+    // before when
+    let oldInvoice = db.invoices.get(testData.invoice._id);
+    expect(oldInvoice?.status).toStrictEqual('draft');
+    expect(oldInvoice?.items.length).toStrictEqual(2);
+
+    // when
+    await chargeInvoices();
+
+    // then
+    oldInvoice = db.invoices.get(testData.invoice._id);
+    expect(oldInvoice).toBeDefined();
+    expect(oldInvoice?.status).toStrictEqual('pending');
+    expect(oldInvoice?.date).toStrictEqual(invoiceDate.toDate());
+    expect(oldInvoice?.items.length).toStrictEqual(5);
+    expect(oldInvoice?.totalAmount).toStrictEqual(405.61);
+
+    expect(db.payments.size).toBe(1);
+    const payment = Array.from(db.payments.values())[0];
+    expect(payment).toBeDefined();
+    expect(payment.status).toStrictEqual('pending');
+    expect(payment.amount).toStrictEqual(405.61);
+
+    expect(db.invoices.size).toBe(2);
+    let newInvoice = Array.from(db.invoices.values()).at(-1);
+    expect(newInvoice).toBeDefined();
+    expect(newInvoice?.date).toStrictEqual(invoiceDate.add(1, 'month').toDate());
```

Comment:
@mariusheine Could you maybe help me with this? There is an initial invoice created for `2020-01-31T23:00:00.000Z` which is the end of the first period of the test subscription. Now I iterate the day after the invoice date over the subscription, charge the open invoice by creating a payment and create a new invoice. The new invoice however is not getting the correct date IMO. Is this in general correct?

Reply:
@anbraten I will have  a look

Reply:
@mariusheine Thanks.

Reply:
@anbraten Ok I also think the period is wrong. It is because you call `getPeriodFromAnchorDate` with the 1st of february. This indeed is in the period 01.02 til 28.02. 

I think the easiest would be to put an adjusted invoice date into `getPeriodFromAnchorData` at the top of `addSubscriptionChangesToInvoice`. 

Or we extend the invoice model by storing the invoice data next to the period information. Invoice date would be the date when the invoice has been generated and period information would be the billing period.

Reply:
@anbraten I adjusted it in [58efac4](https://github.com/geprog/gringotts/pull/39/commits/58efac40bfa5fc69d5af9f218d111020270c31c7)

The expected invoice amounts did not fit so I simply updated them. Have you calculated the values by your own? Neverthless we should recheck manually if the amount is correct now.